### DISCUSSION TITLE
✨ allow to set the container runtime via config/env var

### DISCUSSION THREAD

**Discussion by c33s** [dagger/dagger#12132] [Ideas]
### What are you trying to do?

according the guide to use [podman](https://docs.dagger.io/integrations/podman/) you should symlink podman to docker. 

it would be much better to allow to configure the "runtime" via config and/or environment variable.  something like `DAGGER_RUNTIME=/usr/bin/docker` or `DAGGER_RUNTIME=/usr/bin/podman`

(btw the issue template is quite uncool for keyboard first persons if you have to jump between the three blocks "what are you trying to do", "why is this important to you" and "how are you currently working around this?")

### Why is this important to you?

as i am using podman and docker in parallel because i am testing images against both "runtimes" and also have trouble to fully switch to podman as sometimes docker is a hard requirement.

### How are you currently working around this?

~~script to rename `docker` to `docker_disabled` and create the symlink from `docker -> podman` which of course needs sudo -.-~~
`.envrc` for [direnv](https://github.com/direnv/direnv)
```
ln -sf $(which podman) $HOME/.local/bin-tmp/docker
export PATH=$HOME/.local/bin-tmp:$PATH
```

**Comment by gerhard** [comment]
Thank your for all the context @c33s.

While I don't use Podman myself, I am wondering whether you could leverage `DOCKER_HOST` to point to e.g. `unix:///run/user/1000/podman/podman.sock`

With the help of `direnv`, the projects that need this combo could use e.g. `export DOCKER_HOST=unix:///run/user/1000/podman/podman.sock` . Using this config, Dagger would still use the `docker` CLI default for managing the Engine, **but** the `docker` CLI would be connecting to Podman instead of Docker. You can still keep both container runtimes installed, and select which to use on a per path basis.

If it helps, these are the functions which are available in my shell for quickly switching between different container runtimes for Dagger:

```sh
# use remote Linux Docker Engine on x22 - NixOS
dx22() {
  export DOCKER_HOST=tcp://100.81.87.121:2375
}

# use remote Linux Docker Engine on h22 - Pop!_OS
dh22() {
  dh22lan
}
dh22lan() {
  export DOCKER_HOST=tcp://192.168.88.22:2375
}
dh22wan() {
  export DOCKER_HOST=tcp://100.110.157.107:2375
}
dh22ssh() {
  export DOCKER_HOST=ssh://h22
}
# DEFAULT to dh22
dh22wan

# use remote Linux Docker Engine on h23 - Pop!_OS
dh23() {
  dh23lan
}
dh23lan() {
  export DOCKER_HOST=tcp://192.168.88.23:2375
}
dh23wan() {
  export DOCKER_HOST=tcp://100.83.148.49:2375
}

# use remote Linux Docker Engine on h24 - Ubuntu
dh24() {
  dh24lan
}
dh24lan() {
  export DOCKER_HOST=tcp://192.168.88.24:2375
}
```

Notice how I mix WAN (Tailscale for me), LAN & SSH connectivity depending on what I'm testing. I have multiple Docker instances for redundancy purposes, but also for testing differences between versions / kernels. Maybe this is helpful for what you are trying to achieve.

I also want to make sure that you are aware of our `_EXPERIMENTAL_DAGGER_RUNNER_HOST=podman-container://...` config that bypasses any auto provisioning and just uses an existing Dagger Engine running as a Podman container: https://docs.dagger.io/configuration/custom-runner

<img width="854" alt="image" src="https://github.com/user-attachments/assets/bbf51b0a-790b-4741-8236-f068581b0aba" />

There are a few reasons why adding the `DAGGER_RUNTIME` environment variable is not a good idea from my perspective:
- It adds another configuration knob (the maintenance burden is significant)
- One day we might get to tackling this properly: https://github.com/dagger/dagger/issues/5583 🤔
- [Here is an example of a Dagger module that provisions Dagger Engines on remote platforms](https://github.com/gerhard/daggerverse/tree/7f47b54e2dd8daa18a38fc79246e2fd5f93a7f05/dagr). Maybe we can expand this to encode managing, configuring & using Dagger Engines on Podman 🤷‍♂️

Did the above help @c33s ?
